File: src/leetcode/0006-ZigZag-Conversion/solution.cpp
// Copyright (c) 2019 shaqsnake. All rights reserved.
/**
 * @Author: shaqsnake
 * @Email: shaqsnake@gmail.com
 * @Date: 2019-08-13 10:23:26
 * @LastEditTime: 2019-08-13 14:26:48
 * @Description: 6. ZigZag Conversion
 */

class Solution {
public:
    string convert(string s, int n) {
        if (n == 1) return s;

        string res;
        for (int i = 0; i < n; i++) {
            if (i == 0 || i == n - 1) {
                for (int j = i; j < s.size(); j += 2 * (n - 1)) res += s[j];
            } else {
                for (int j = i, k = 2 * (n - 1) - i; j < s.size() || k < s.size(); j += 2 * (n - 1), k += 2 * (n - 1)) {
                    if (j < s.size()) res += s[j];
                    if (k < s.size()) res += s[k];
                }
            }
        }

        return res;
    }
};

File: src/leetcode/0006-ZigZag-Conversion/solution.py
'''
@Author: shaqsnake
@Email: shaqsnake@gmail.com
@Date: 2019-05-09 13:47:55
@LastEditTime: 2019-08-13 10:26:10
@Description: 6. ZigZag Conversion
'''


class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        # 边界条件
        if numRows <= 1 or len(s) < numRows:
            return s

        # 先根据numRows的大小n构建大小n的List，用来缓存每一行的字符
        L = [''] * numRows
        cur = step = 0
        # 遍历s，把s中的字符依次放入List，在抵达边界时折返
        for x in s:
            # 把字符逐一放入对应的List
            L[cur] += x
            # 当指针抵达上届时折返
            if cur == 0:
                step = 1
            # 当指针抵达下界时折返
            elif cur == numRows - 1:
                step = -1
            cur += step
        return ('').join(L)
